gobblin-runtime/src/main/java/org/apache/gobblin/runtime/kafka/HighLevelConsumer.java

```diff
@@ -125,81 +181,127 @@ protected void createMetrics() {
   }
 
   /**
-   * Called every time a message is read from the stream. Implementation must be thread-safe if {@link #numThreads} is
+   * Called every time a message is read from the queue. Implementation must be thread-safe if {@link #numThreads} is
    * set larger than 1.
    */
-  protected abstract void processMessage(MessageAndMetadata<K, V> message);
+  protected abstract void processMessage(DecodeableKafkaRecord<K,V> message);
 
   @Override
   protected void startUp() {
     buildMetricsContextAndMetrics();
-    this.consumer = createConsumerConnector();
+    processQueues();
+    if(!enableAutoCommit) {
+      offsetCommitExecutor = Executors.newSingleThreadScheduledExecutor();
+      offsetCommitExecutor.scheduleAtFixedRate(new Runnable() {
+        @Override
+        public void run() {
+          offsetCommitter();
+        }
+      }, 0,100, TimeUnit.MILLISECONDS);
+    }
 
-    List<KafkaStream<byte[], byte[]>> streams = createStreams();
-    this.executor = Executors.newFixedThreadPool(this.numThreads);
+    mainExecutor.scheduleAtFixedRate(new Runnable() {
+      @Override
+      public void run() {
+        consume();
+      }
+    }, 0, 50, TimeUnit.MILLISECONDS);
+  }
 
-    // now create an object to consume the messages
-    //
-    int threadNumber = 0;
-    for (final KafkaStream stream : streams) {
-      this.executor.execute(new MonitorConsumer(stream));
-      threadNumber++;
+  /**
+   * Consumes {@link KafkaConsumerRecord}s and adds to a queue
+   * Note: All records from a KafkaPartition are added to the same queue.
+   * A queue can contain records from multiple partitions if partitions > numThreads(queues)
+   */
+  private void consume() {
+    try {
+      Iterator<KafkaConsumerRecord> itr = gobblinKafkaConsumerClient.consume();
+      while (itr.hasNext()) {
+        KafkaConsumerRecord record = itr.next();
+        int idx = record.getPartition() % numThreads;
+        queues[idx].put(record);
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
     }
   }
 
-  protected ConsumerConfig createConsumerConfig(Config config) {
-    Properties props = ConfigUtils.configToProperties(config);
 
-    if (!props.containsKey(GROUP_ID_KEY)) {
-      props.setProperty(GROUP_ID_KEY, DEFAULT_GROUP_ID);
+  private void processQueues() {
+    for(BlockingQueue queue : queues) {
+      queueExecutor.execute(new QueueProcessor(queue));
     }
-    return new ConsumerConfig(props);
   }
 
-  protected ConsumerConnector createConsumerConnector() {
-    return Consumer.createJavaConsumerConnector(this.consumerConfig);
+  /**
+   * A thread that commits offsets to kafka at regular intervals
+   */
+  private void offsetCommitter() {
+    if(recordsProcessed.intValue() >= offsetsCommitNumRecordsThreshold || ((System.currentTimeMillis() - lastCommitTime) / 1000 >= offsetsCommitTimeThresholdSecs)) {
+      Map<KafkaPartition, Long> copy = new HashMap<>(partitionOffsetsToCommit);
```

Comment:
Why do we need to make a copy? And why clear the map on every commit? What happens if a thread writes offsets to partitionOffsetsToCommit between line 246 and 247? Won't this cause some offsets to not be committed?